Mobile/ShopLens/Droid/Helpers/ShopLensTextToSpeech.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using Android.Content;
+using Android.Runtime;
+using Android.Speech.Tts;
+using Java.Util;
+
+namespace ShopLens.Droid.Helpers
+{
+    class ShopLensTextToSpeech : Java.Lang.Object, TextToSpeech.IOnInitListener
+    {
+        event EventHandler TTSInitialized;
+        ShopLensUtteranceProgressListener uttListener;
+
+        public TextToSpeech tts;
+
+        public ShopLensTextToSpeech(Context context, Action<object, EventArgs> ReactToTTSInit,
+            Action<object, UtteranceIdArgs> ReactToEndOfSpeech)
+        {
+            tts = new TextToSpeech(context, this);
+            uttListener = new ShopLensUtteranceProgressListener(ReactToEndOfSpeech);
+
+            tts.SetOnUtteranceProgressListener(uttListener);
+            TTSInitialized += (obj, eargs) => ReactToTTSInit(obj, eargs);
+        }
+
+        // TTS Engine method called when TTS is initialized.
+        public void OnInit([GeneratedEnum] OperationResult status)
+        {
+            if (status == OperationResult.Error)
+            {
+                tts.SetLanguage(Locale.Default);
+            }
+            else if (status == OperationResult.Success)
+            {
+                tts.SetLanguage(Locale.English);
+            }
+
+            TTSInitialized(this, new EventArgs());
+        }
+
+        public void Speak(string message, string utteranceId)
+        {
+            tts?.Speak(message, QueueMode.Flush, null, utteranceId);
```

Comment:
Right now this is a potential silent failure. 

Is this the defined behavior?

See line 48 as well.

Reply:
Should we throw an exception on line 43 then?
Line 48 is good, as in, if Tts isn't initialized, we don't need to call stop. 
You could write ?. in Main activity in the OnStop method, but then the Main Activity needs to know that this class's method Stop can't be called when tts is null. It's null when TalkBack is null, so in order for Main Activity to function properly it would need to know, see inside this class. So if this class has a null check then Main doesn't need to know anything about its internals.